test_conformance/extensions/cl_khr_external_memory_ahb/test_ahb.cpp

```diff
@@ -0,0 +1,1899 @@
+
+        AHardwareBuffer_Desc aHardwareBufferDesc = { 0 };
+        aHardwareBufferDesc.format = format.aHardwareBufferFormat;
+        for (auto usage : test_usages)
+        {
+            // Filter out usage flags that are not readable on device
+            if (!isAHBUsageReadable(usage.usageFlags))
+            {
+                continue;
+            }
+
+            aHardwareBufferDesc.usage = usage.usageFlags;
+            for (auto resolution : test_sizes)
+            {
+                aHardwareBufferDesc.width = resolution.width;
+                aHardwareBufferDesc.height = resolution.height;
+                aHardwareBufferDesc.layers = 1;
+
+                CHECK_AHARDWARE_BUFFER_SUPPORT(aHardwareBufferDesc, format);
+
+                AHardwareBuffer *aHardwareBuffer = nullptr;
+                int ahb_result = AHardwareBuffer_allocate(&aHardwareBufferDesc,
+                                                          &aHardwareBuffer);
+                if (ahb_result != 0)
+                {
+                    log_error("AHardwareBuffer_allocate failed with code %d\n",
+                              ahb_result);
+                    return TEST_FAIL;
+                }
+
+                // Determine AHB memory layout
+                AHardwareBuffer_Desc hardware_buffer_desc = {};
+                AHardwareBuffer_describe(aHardwareBuffer,
+                                         &hardware_buffer_desc);
+
+                test_assert_error(hardware_buffer_desc.width
+                                      == resolution.width,
+                                  "AHB has unexpected width");
+                test_assert_error(hardware_buffer_desc.height
+                                      == resolution.height,
+                                  "AHB has unexpected height");
+
+                // Populate AHB with random data
+                size_t pixelSize = get_pixel_size(&format.clImageFormat);
+                image_descriptor imageInfo = { 0 };
+                imageInfo.format = &format.clImageFormat;
+                imageInfo.type = format.clMemObjectType;
+                imageInfo.width = resolution.width;
+                imageInfo.height = resolution.height;
+                imageInfo.rowPitch = hardware_buffer_desc.stride * pixelSize;
+                test_assert_error(imageInfo.rowPitch
+                                      >= pixelSize * imageInfo.width,
+                                  "Row pitch is smaller than width");
+
+                size_t srcBytes = get_image_size(&imageInfo);
+                test_assert_error(srcBytes > 0, "Image cannot have zero size");
+
+                BufferOwningPtr<char> srcData;
+                generate_random_image_data(&imageInfo, srcData, seed);
+
+                void *hardware_buffer_data = nullptr;
+                ahb_result = AHardwareBuffer_lock(
+                    aHardwareBuffer, AHARDWAREBUFFER_USAGE_CPU_WRITE_OFTEN, -1,
```

Comment:
There is a mismatch here between `AHARDWAREBUFFER_USAGE_CPU_WRITE_OFTEN` and `usage` that was used to create the AHB.

Some of the flags in `test_usages` don't contain any CPU flags and so the call to `AHardwareBuffer_lock` can fail in those cases.

I think all of the variants in `test_usages` will need CPU flags set on them.